src/int/specialized_div_rem/trifecta.rs

```diff
@@ -0,0 +1,453 @@
+macro_rules! impl_trifecta {
+    (
+        $unsigned_name:ident, // name of the unsigned function
+        $signed_name:ident, // name of the signed function
+        $half_division:ident, // function for division of a $uX by a $uX
+        $n_h:expr, // the number of bits in $iH or $uH
+        $uH:ident, // unsigned integer with half the bit width of $uX
+        $uX:ident, // the largest division instruction that this function calls operates on this
+        $uD:ident, // unsigned integer with double the bit width of $uX
+        $iD:ident, // signed version of $uD
+        $($unsigned_attr:meta),*; // attributes for the unsigned function
+        $($signed_attr:meta),* // attributes for the signed function
+    ) => {
+        /// Computes the quotient and remainder of `duo` divided by `div` and returns them as a
+        /// tuple.
+        /// 
+        /// This is optimized for division of two integers with bit widths twice as large
+        /// as the largest hardware integer division supported. Note that some architectures supply
+        /// a division of an integer larger than register size by a regular sized integer (e.x.
+        /// x86_64 has a `divq` instruction which can divide a 128 bit integer by a 64 bit integer).
+        /// In that case, the `_asymmetric` algorithm should be used instead of this one.
+        /// 
+        /// This is called the trifecta algorithm because it uses three main algorithms: short
+        /// division for small divisors, the "mul or mul - 1" algorithm for when the divisor is
+        /// large enough for the quotient to be determined to be one of two values via only one
+        /// small division, and an underguessing long division algorithm for middle cases.
+        ///
+        /// # Panics
+        ///
+        /// When attempting to divide by zero, this function will panic.
+        $(
+            #[$unsigned_attr]
+        )*
+        pub fn $unsigned_name(duo: $uD, div: $uD) -> ($uD,$uD) {
+            // wrapping_{} was used everywhere for better performance when compiling with
+            // `debug-assertions = true`
+
+            // This replicates `carrying_mul` (rust-lang rfc #2417). LLVM correctly optimizes this
+            // to use a widening multiply to 128 bits.
```

Comment:
But cg_clif won't, causing an infinite loop. :(

Reply:
Also is it optimized correctly on archs other than x86?

Reply:
I presume `cg_clif` is the cranelift backend. I would understand that cranelift can't optimize it into a single instruction, but what do you mean by infinite loop?

Reply:
Never mind, thought this function was for multiplication, in which case the `wrapping_mul` would end up calling this function again.

> I presume cg_clif is the cranelift backend.

Indeed